### ISSUE TITLE
Neptune-Core: Stack Trace on Shutdown

### ISSUE THREAD

**Issue by aszepieniec** [Neptune-Crypto/neptune-core#236]
After starting `neptune-core` with `./scripts/linux/run-single-instance`, running it for about 15s, and then terminating with Ctrl-C. This happens on my laptop which has 12 threads and 31 GB of RAM.

```
^CI0:  main_loop: Detected Ctrl+c signal.
I0:  main_loop: Received SIGINT
I0:  main_loop: Shutdown initiated.
I0:  main_loop: sent bye
alan@olivas:~/repos/neptune-core$ I0:  models::state: Flushed all databases
I0:  main_loop: Shutdown completed.
I0:  job_queue::queue: JobQueue: received stop message.  stopping.
thread 'tokio-runtime-worker' panicked at src/job_queue/queue.rs:165:48:
called `Result::unwrap()` on an `Err` value: SendError { .. }
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fbab78289dd8c6e8860034e0048cfb538f217700/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/fbab78289dd8c6e8860034e0048cfb538f217700/library/core/src/panicking.rs:75:14
   2: core::result::unwrap_failed
             at /rustc/fbab78289dd8c6e8860034e0048cfb538f217700/library/core/src/result.rs:1698:5
   3: core::result::Result<T,E>::unwrap
             at /home/alan/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1103:23
I0:  job_queue::queue::JobQueue<P>::start::{{closure}}
             at ./src/job_queue/queue.rs:165:29
   5: <tokio_util::task::task_tracker::TrackedFuture<F> as core::future::future::Future>::poll
             at /home/alan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-util-0.7.11/src/task/task_tracker.rs:669:9
   6: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/alan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.40/src/instrument.rs:321:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by aszepieniec** [comment]
After some digging help from @Sword-Smith:
 - The stack trace is generated on line 165 of `src/job_queue/queue.rs` by the call to `unwrap()`.
 - No stack trace is generated when the max log2 padded height is set to 15.
 - We think it is related to line 137.. in `src/models/proof_abstractions/tasm/prover_job.rs` that deals with what happens if proof complexity is exceeded `return Err(ProverJobError::ProofComplexityLimitExceeded { [...]`.

Handing off to @dan-da since he knows the architecture of the proof queue better and since he probably already has 85% of a solution lying around somewhere.

**Comment by Sword-Smith** [comment]
Here's another stack trace I saw when shutting down the client with `ctrl + c` while composing:

```
I0:  peer_loop: Responding with: [([::ffff:127.0.0.1]:29791, 337599510772526109330102094169730634519), ([::ffff:127.0.0.1]:29792, 234152638593612143812170491281724457518)]
I2:  main_loop: Attempting to run transaction-proof-upgrade
I2:  main_loop: Found no transaction-proof to upgrade
I2:  peer_loop: Received peer list resp from peer [::ffff:127.0.0.1]:59618
I2:  main_loop: Received message sent to main task.
I2:  main_loop: Received peer discovery answer from a peer task
^CI2:  main_loop: Received SIGINT
I0:  main_loop: Received SIGINT
I2:  main_loop: Detected Ctrl+c signal.
I0:  main_loop: Detected Ctrl+c signal.
I2:  main_loop: Shutdown initiated.
I2:  main_loop: sent bye
I0:  main_loop: Shutdown initiated.
I0:  main_loop: sent bye
I2:  peer_loop: Handling disconnect all message from main in peer loop
I0:  peer_loop: Handling disconnect all message from main in peer loop
I0:  peer_loop: handle_main_task_message is closing the connection to [::ffff:127.0.0.1]:56168
I0:  peer_loop: Handling disconnect all message from main in peer loop
I0:  peer_loop: handle_main_task_message is closing the connection to [::ffff:127.0.0.1]:29792
I2:  peer_loop: handle_main_task_message is closing the connection to 127.0.0.1:29791
I0:  peer_loop: Exited peer loop for [::ffff:127.0.0.1]:56168
I2:  peer_loop: Handling disconnect all message from main in peer loop
I2:  peer_loop: handle_main_task_message is closing the connection to [::ffff:127.0.0.1]:59618
I2:  peer_loop: Exited peer loop for 127.0.0.1:29791
I0:  peer_loop: Exited peer loop for [::ffff:127.0.0.1]:29792
I1:  main_loop: Received SIGINT
I2:  peer_loop: Exited peer loop for [::ffff:127.0.0.1]:59618
thv@mjolnir:~/repos/neptune-core$ I0:  job_queue::queue:   *** JobQueue: ended job #58 - 472.93497 secs ***
I0:  models::state: Flushed all databases
I0:  connect_to_peers: Fetched peer info standing 0 for peer [::ffff:127.0.0.1]:56168
I2:  models::state: Flushed all databases
I2:  connect_to_peers: Fetched peer info standing 0 for peer 127.0.0.1:29791
I0:  connect_to_peers: Stored peer info standing 0 for peer [::ffff:127.0.0.1]:56168
I0:  peer_loop: Ending peer loop for [::ffff:127.0.0.1]:56168
I0:  connect_to_peers: Fetched peer info standing 0 for peer [::ffff:127.0.0.1]:29792
I0:  connect_to_peers: Stored peer info standing 0 for peer [::ffff:127.0.0.1]:29792
I0:  peer_loop: Ending peer loop for [::ffff:127.0.0.1]:29792
I0:  connect_to_peers: Connection closing
I2:  connect_to_peers: Stored peer info standing 0 for peer 127.0.0.1:29791
I2:  peer_loop: Ending peer loop for 127.0.0.1:29791
I2:  connect_to_peers: Connection closing
I2:  connect_to_peers: Fetched peer info standing 51 for peer [::ffff:127.0.0.1]:59618
I2:  connect_to_peers: Stored peer info standing 51 for peer [::ffff:127.0.0.1]:59618
I2:  peer_loop: Ending peer loop for [::ffff:127.0.0.1]:59618
I2:  main_loop: Shutdown completed.
I2:  job_queue::queue: JobQueue: received stop message.  stopping.
thread 'tokio-runtime-worker' panicked at /home/thv/repos/neptune-core/src/mine_loop.rs:517:26:
unexpected error while composing: external proving process failed

Caused by:
    proving process did not return any exit code

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
             at /home/thv/.cargo/registry/src/index.crates.io-6f17d22bba15001f/anyhow-1.0.86/src/backtrace.rs:27:14
   1: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /home/thv/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:2008:27
I0:  models::proof_abstractions::tasm::program::prove_consensus_program::{{closure}}
             at ./src/models/proof_abstractions/tasm/program.rs:187:8
I0:  models::proof_abstractions::tasm::program::ConsensusProgram::prove::{{closure}}
             at ./src/models/proof_abstractions/tasm/program.rs:141:14
I0:  models::blockchain::transaction::validity::single_proof::SingleProof::produce::{{closure}}
             at ./src/models/blockchain/transaction/validity/single_proof.rs:301:14
stack backtrace:
   0: rust_begin_unwind
             at /rustc/3ed6e3cc69857129c1d314daec00119ff47986ed/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/3ed6e3cc69857129c1d314daec00119ff47986ed/library/core/src/panicking.rs:74:14
I0:  mine_loop::mine::{{closure}}
             at ./src/mine_loop.rs:517:26
I0:  initialize::{{closure}}::{{closure}}
             at ./src/lib.rs:236:18
   4: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /home/thv/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/future.rs:123:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by dan-da** [comment]
The reported issue should be fixed in aa009d0dab1f529ccdbdae2f05d4e623ef3cbdb9.

**Comment by dan-da** [comment]
> Here's another stack trace I saw when shutting down the client with `ctrl + c` while composing:
> 
> ```
> unexpected error while composing: external proving process failed
> 
> Caused by:
>     proving process did not return any exit code
> ```

this is a different issue.  I will keep an eye out for it, and think about it.

note that mine_loop now explicitly checks for proving errors and:
 1. pauses mining if the error is ProofComplexityLimitExceeded
 2. intentionally panics if the error is anything else.  (which is happening here)

it would be a simple change, and perhaps the right thing, to just pause mining loop for any proving error.   As one example, a user could have intentionally killed the triton-vm-prover.  Should that end neptune-core or not?

Unfortunately there does not appear to be any reliable cross platform way to detect if a child process was terminated or exited on its own.  So what we end up with is that it either exited with 0, with non-zero, or with no exit code at all.   On unix, the latter generally means it was killed via a signal.  But on windows a kill may be one of various exit codes.

anyway, I'm closing this.  we can open another issue for the 2nd stack trace if need be.
